Implement sample variant annotation dataflow pipeline

calbach opened the pull request:
See various caveats and disclaimers in comments: this is a limited sample application.

One thing which may need revision is the output; right now it's really only human readable (at best). Open to suggestions on a better output format.

calbach commented on src/main/java/com/google/cloud/genomics/dataflow/utils/GenomicsDatasetOptions.java:
```diff
@@ -63,37 +63,49 @@ public static void validateOptions(GenomicsDatasetOptions options) {
+  @Description("The IDs of the Google Genomics call sets this pipeline is working with, comma "
+      + "delimited.Defaults to 1000 Genomes HG00261.")
+  @Default.String("10473108253681171589-0")
+  String getCallSetIds();
```
@deflaux let me know if you'd prefer I fork into my own options at this point. Not sure how much we want to jam into this object.

deflaux replied:
I think this is fine place for it for now.

deflaux commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/AnnotateVariants.java:
```diff
@@ -0,0 +1,307 @@
+              .setPageSize(VARIANTS_PAGE_SIZE),
+            VARIANT_FIELDS))
+        .filter(VariantUtils.IS_SNP);
+    for (Variant variant : varIter) {
```
To make sure I'm understanding this correctly, are these assumptions true?

*  By default, this pipeline will yield an output record for every alternate allele in 1,000 Genomes within BRCA1 that is a SNP and has an effect other than synonymous.
*  For 1,000 genomes, restricting to sample HG00261 has no bearing on the output of this pipeline since all samples have calls for all variants (and we are also not retrieving/looking at the genotype within the call).
* If we change the job parameters to run on Platinum Genomes and a callSetId within it, we will only annotate the variants that the specified callSetId has.

calbach replied:
That's correct. People typical run a variant annotation program on a single VCF, so I think the behavior is reasonably well aligned with a user's expectations.

pgrosu replied:
CH, is right but you still want to keep track of what you're annotating since metadata is still important if you combine datasets or are comparing them.  If you can cache them that will save you time later on.

deflaux commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/AnnotateVariants.java:
```diff
@@ -0,0 +1,307 @@
+                .setType("SNP")
+                .setEffect(effect.toString())
+                .setGeneId(transcript.getTranscript().getGeneId())
+                .setTranscriptIds(ImmutableList.of(transcript.getId()))));
```
Is this output sufficient to know which particular SNP had the effect that was other than synonymous?

pgrosu replied:
This is just the very beginning, and there are a lot more opportunities to expand on here.  For instance, multiple sites are now currently ignored in [line 128 of AnnotationUtils.java](https://github.com/googlegenomics/dataflow-java/blob/variant-annotation/src/main/java/com/google/cloud/genomics/dataflow/utils/AnnotationUtils.java#L128):

```
if (codonExonOffset < 0 || exonBases.length() <= codonExonOffset+3) {
  LOG.fine("variant codon spans multiple exons, this case is not yet handled");
  return null;
}
```

There are SIFT and PolyPhen prediction to take into account.  Another option is to filter by minor allele frequency (MAF).  Again many other options to take into account to be sure the search is not too stringent, and yet as complete as possible for the dataset(s) and analysis being pursued.  

What about filtering on annotated subsets of data you generated and you deem they should be excluded as collection (set), based on analysis you generated?  This is a huge area, which is an understatement.

Paul

calbach replied:
@deflaux it previously emitted just the variantId. I've modified the output to now also include the position and allele.

deflaux commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/AnnotateVariants.java:
```diff
@@ -0,0 +1,307 @@
+import java.util.logging.Logger;
+
+/**
+ * Demonstrates a simple variant annotation program which takes
```
This is a _very_ cool sample!!!

deflaux commented:
This looks good to me - merge it at your convenience.

garrickevans commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/AnnotateVariants.java:
```diff
@@ -0,0 +1,307 @@
+
+    Pipeline p = Pipeline.create(opts);
+    DataflowWorkarounds.registerGenomicsCoders(p);
+    DataflowWorkarounds.getPCollection(ImmutableList.copyOf(contigs), p, opts.getNumWorkers())
```
Is this necessary? Why not convert the list of Contigs to a PCollection directly?

calbach replied:
Done

garrickevans commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/AnnotateVariants.java:
```diff
@@ -0,0 +1,307 @@
+    int varCount = 0;
+    Iterable<Variant> varIter = FluentIterable
+        .from(Paginator.Variants.create(genomics, ShardBoundary.STRICT).search(
+            c.element().getVariantsRequest(varsetId)
```
s/c.element()/contig

calbach replied:
Done

calbach commented:
Rebased, made some performance changes, and added some timing information. The end result is that it will currently work well on small regions, but performs quite poorly on whole variant-sets, on account of SearchVariants throughput. This should improve over time.

deflaux commented:
Nice sample CH!
